doc/howtos/rdtraining/constraints.rst

```diff
@@ -0,0 +1,128 @@
+
+    - A property expected price must be strictly positive
+    - A property selling price must be positive
+    - An offer price must be strictly positive
+    - A property tag name and property type name must be unique
+
+    Tip: search for the ``unique`` keyword in the Odoo codebase for the name uniqueness.
+
+Restart the server with the ``-u estate`` option to see the result. Note that you might have data
+preventing an SQL constraint to be set. such error message might pop up.
+
+.. code-block:: text
+
+    ERROR rd-demo odoo.schema: Table 'estate_property_offer': unable to add constraint 'estate_property_offer_check_price' as CHECK(price > 0)
+
+For example, if some offers have a price of zero, the constraint can't be applied. You can delete
+the problematic data in order to apply the new constraints.
+
+Python
+======
+
+**Reference**: the documentation related to this topic can be found in
+:func:`~odoo.api.constrains`.
+
+.. note::
+
+    **Goal**: at the end of this section, it is not possible to accept an offer with a price
+    lower than 90% of the expected price.
+
+    .. image:: constraints/media/python.gif
+        :align: center
+        :alt: Python constraint
+
+SQL constraints are an efficient way of ensuring data consistency. However, such constrains cannot
+be 'dynamic' in the sense that cross-checking data might not be possible. In our real estate module
+we would like to prevent a selling price lower than 90% of the expected price. To do this, we
+need a Python constraint.
+
+A Python constraint is defined as a method decorated with
+:func:`~odoo.api.constrains`, and invoked on a recordset. The decorator
+specifies which fields are involved in the constraint, so that the constraint is
+automatically evaluated when one of them is modified. The method is expected to
+raise an exception if its invariant is not satisfied::
+
+    from odoo.exceptions import ValidationError
+
+    ...
+
+    @api.constrains('age')
+    def _check_something(self):
+        for record in self:
+            if record.age > 20:
+                raise ValidationError("Your record is too old: %s" % record.age)
+        # all records passed the test, don't return anything
+
+A simple example can be found
+`here <https://github.com/odoo/odoo/blob/3783654b87851bdeb11e32da78bb5b62865b869a/addons/account/models/account_payment_term.py#L104-L108>`__.
+
+.. exercise:: Add Python constraints
+
+    Add a constraint so that the selling price cannot be lower than 90% of the expected price.
+
+    Tip: the selling price is zero until an offer is validated. You will need to fine tune your
+    check to take this into account.
+
+    .. warning::
+
+        Always use the :meth:`~odoo.tools.float_utils.float_compare` and
+        :meth:`~odoo.tools.float_utils.float_is_zero` methods when comparing floats!
+
+Be sure the constraint is triggered every time the selling price or the expected price is changed!
+
```

Comment:
As you can use Python constraints to do the same Job as SQL constraints, maybe explain why you need to choose one over the other. Correct me if I'm wrong, but for performances issues is better to use (when it's possible) SQL constraints non ?